tests/test_implicit_networks.cpp

```diff
@@ -187,20 +200,190 @@ TEST_CASE("implicit arrangement on known examples", "[IA][examples]") {
+        size_t func_id;
+        
+        if (!load_functions(std::string(TEST_FILE) + "/3-sphere-3.json", pts, funcVals)) {
+            throw std::runtime_error("ERROR: Failed to load functions.");
+        }
+
+        // compute implicit arrangement
+        bool success = implicit_arrangement(
+                robust_test,
+                use_lookup,
+                use_secondary_lookup,
+                use_topo_ray_shooting,
+                //
+                pts, tets, funcVals,
+                //
+                iso_pts,iso_faces,patches,
+                patch_function_label,
+                iso_edges,chains,
+                non_manifold_edges_of_vert,
+                shells,arrangement_cells,cell_function_label,
+                timing_labels,timings,
+                stats_labels,stats);
+        REQUIRE(success);
+
+        // check
+        REQUIRE(patches.size() == 12);
+        REQUIRE(patch_function_label.size() == 12);
+        REQUIRE(chains.size() == 6);
+        REQUIRE(arrangement_cells.size() == 8);
+        
+        // function label check
+        std::vector<size_t> patch_gt = {0, 2, 0, 2, 1, 1, 2, 1, 0, 1, 2, 0};
+        REQUIRE(patch_function_label == patch_gt);
+        std::vector<std::vector<size_t>> cell_gt = {{1, 0, 0},{0, 0, 0}, {0, 0, 1}, {1, 0, 1}, {1, 1, 0}, {1, 1, 1}, {0, 1, 1}, {0, 1, 0}};
+        REQUIRE(cell_function_label == cell_gt);
+    }
+    
+    SECTION("three spheres config 4") {
+        // compute function values on tet grid vertices
+        size_t n_pts = pts.size();
+        size_t n_func = 3;
+        funcVals.resize(n_pts, n_func);
+        size_t func_id;
+        
+        if (!load_functions(std::string(TEST_FILE) + "/3-sphere-4.json", pts, funcVals)) {
+            throw std::runtime_error("ERROR: Failed to load functions.");
+        }
+
+        // compute implicit arrangement
+        bool success = implicit_arrangement(
+                robust_test,
+                use_lookup,
+                use_secondary_lookup,
+                use_topo_ray_shooting,
+                //
+                pts, tets, funcVals,
+                //
+                iso_pts,iso_faces,patches,
+                patch_function_label,
+                iso_edges,chains,
+                non_manifold_edges_of_vert,
+                shells,arrangement_cells,cell_function_label,
+                timing_labels,timings,
+                stats_labels,stats);
+        REQUIRE(success);
+
+        // check
+        REQUIRE(patches.size() == 3);
+        REQUIRE(patch_function_label.size() == 3);
+        REQUIRE(chains.size() == 0);
+        REQUIRE(arrangement_cells.size() == 4);
+        
+        // function label check
+        std::vector<size_t> patch_gt = {2, 1, 0};
+        REQUIRE(patch_function_label == patch_gt);
+        std::vector<std::vector<size_t>> cell_gt = {{0, 0, 1}, {0, 0, 0}, {0, 1, 1}, {1, 1, 1}};
+        REQUIRE(cell_function_label == cell_gt);
     }
```

Comment:
Can you add another 3 sphere unit test: The first sphere has very large radius such that its isosurface does not intersect the grid. The second sphere and third sphere are regular sized sphere intersecting each other. The second sphere has normal orienting outward, and the third sphere has normal orienting inward.

Reply:
> The function could have an oversized shape that are not captured by the bounding, like this example, or a highly curved shape such that the grid is too coarse to capture the function. The easy solution to this issue is that once we find a function that its sign is not activated by any of the patches, we can propagate its sign at any point on the grid to all the cells. Would that be too hand-wavy? Or is there a better solution to this issue?

Yes, this test case is designed to check this situation. If an isosurface does not touch the grid at all, we can assume it has the same sign for all cells based on our assumptions. Since we already have the sign of the function at grid points, we can use that as the sign for all cells.